### DISCUSSION TITLE
Ability to disable "Displays have separate Spaces"

### DISCUSSION THREAD

**Discussion by IonelLupu** [asmvik/yabai#1553] [Q&A] [answered]
Why isn't it possible to make yabai work with "Displays have separate Spaces" off? What's blocking it? 

The thing is my workflow requires me to have to option disabled. 😢

**Comment by asmvik** [comment, marked as answer]
If I remember correctly this setting alters the behaviour of system APIs that yabai uses to retrieve information. It also requires changes to the scripting-addition that is injected into Dock.app to support the feature-set that is there today.

It probably isn't impossible to support, but I am not interested in spending the amount of time required to do so, because it is not useful to me. If it could be done in a few hours I would do it, but it is unfortunately far more time consuming.

**Reply by IonelLupu** [reply]
Thanks for the info @koekeishiya. Do you know some alternative software that works with MacOS Ventura? I found some alternatives but they didn't update to the latest MacOS version yet :(. yabai is the most up to the app I've found (congrats for that 😄 )

**Reply by asmvik** [reply]
I'm not really up to speed with regards to alternative window management software, sorry. The only one at the top of my head would be Amethyst, but I didn't like the way tiles windows personally, which is why I originally started this series of projects.

**Reply by IonelLupu** [reply]
Oh. I am actually interested only in switching between spaces without any animation and yabai was the only solution. Do you know any other alternatives or solutions to disable the animation when switching spaces to zero?

**Reply by asmvik** [reply]
Uh, no. Currently yabai is the only software for macOS that allows this behaviour, at least to my knowledge. TotalSpaces by BinaryAge used to be able to do this, but as far as I know they stopped maintaining that software some time ago.

However, the specific functionality for disabling native spaces animation could be achieved without having to run yabai, and thus not be affected by the "Displays have separate Spaces" requirement. You'd have to follow the SIP instructions in the wiki, and install the scripting-addition. After the scripting-edition is installed you just need to run `sudo yabai --load-sa`. (Possibly twice). Then just don't run yabai as a service, like the install instructions say you should. 

This should be enough to disable spaces animations, but you will not be able to use `yabai -m space --focus ..` since the actual yabai service is not running. You can however set up and use the macOS provided mission-control keybindings to switch spaces instantaneously at this point. (Cmd+Tab should also bypass space animations).

You would somehow have to automate running `sudo yabai --load-sa` when the Dock.app process restarts (e.g after a logout, system restart, or w/e that could cause the Dock process to re-launch).

**Reply by 24unix** [reply]
I used Total Spaces, it fit my needs in combination with yabai.
My primary use of yabai is the transparency, is there a way to set it just using the sa, without yabai running?
Then I could disable Displays have separate Spaces and have my workflow I want.

**Reply by asmvik** [reply]
Yes, you would need to extract the following function from yabai: https://github.com/koekeishiya/yabai/blob/master/src/sa.m#L482
The more time consuming part would be to create a way to detect and manage open windows, so that you can actually call this function with a target window and modify the opacity.

**Reply by 24unix** [reply]
Thanks for the fast reply. I'll have a look at it.
I already have some Swift code that tracks all windows, maybe I could combine those.

**Comment by IonelLupu** [comment]
> Uh, no. Currently yabai is the only software for macOS that allows this behaviour, at least to my knowledge. TotalSpaces by BinaryAge used to be able to do this, but as far as I know they stopped maintaining that software some time ago.
> 
> However, the specific functionality for disabling native spaces animation could be achieved without having to run yabai, and thus not be affected by the "Displays have separate Spaces" requirement. You'd have to follow the SIP instructions in the wiki, and install the scripting-addition. After the scripting-edition is installed you just need to run `sudo yabai --load-sa`. (Possibly twice). Then just don't run yabai as a service, like the install instructions say you should.
> 
> This should be enough to disable spaces animations, but you will not be able to use `yabai -m space --focus ..` since the actual yabai service is not running. You can however set up and use the macOS provided mission-control keybindings to switch spaces instantaneously at this point. (Cmd+Tab should also bypass space animations).
> 
> You would somehow have to automate running `sudo yabai --load-sa` when the Dock.app process restarts (e.g after a logout, system restart, or w/e that could cause the Dock process to re-launch).

Wow. That's really interesting. Do I have to do this installation from scratch again? I remember I followed the SIP instruction already when I installed yabai a few weeks ago, which means I just have to stop the yabai service, right? Unfortunately, the animation is still there when using macOS's keybdindings 😢   I also run `sudo yabai --load-sa` (twice) and still no luck

**Reply by asmvik** [reply]
I just did a new release v5.0.2 that has Ventura 13.0.0-13.1.0 support. Try to do a full uninstall of yabai (this includes removing the existing entry in accessibility permissions), and then follow the wiki instructions from the beginning. Make sure you don't miss any steps, as macOS is finicky in giving permissions.

**Reply by IonelLupu** [reply]
Alright. I installed the latest version and my old setup works so I will asume everything works correctly (I had to do a full restart on top of what's written in the wiki). But, the macOS shortcut CTRL+Left / CTRL+Right arrows still has an animation. 😞 

What's interesting is that using CTRL+1, CTRL+2 etc to switch Spaces doesn't have an animation (usually, it has a slight animation without yabai). Which is great. (But I don't use those because I have my finger on the CMD key and moving the finger on CTRL+1/2/3 is slow. Using just CMD+LEFT/RIGHT is faster). It seems yabai removed the animation from those shortcuts but not from the CTRL+LEFT/RIGHT shortcuts.  🤔 

Is there something I need to tweak in order to have animations for CTRL+LEFT/RIGHT?

**Reply by asmvik** [reply]
Uh, removing the animation from left/right switches require patching yet another memory location in Dock.app that yabai currently does not do; I actually haven't even tracked down where this is happening, as I've been assuming that users would use yabai, which provides the prev/next functionality. I don't think that I will be trying to track down that part either, sorry.

Maybe I'll copy the spaces manipulation code from yabai and make a separate standalone piece of software that people can run.
Will have to think a bit about how I'd restructure the codebase (without complicating my workflow too much), as they would then share a dependency, and I'm not super interested in maintaining more projects on the side.

**Reply by IonelLupu** [reply]
gotcha. no worries. I really thank you for what you did so far. great job. I think I will try to map the CTRL+1,2,3 to a Karabiner shortcut combination. 

I found this command which presses some keys: `osascript -e 'tell application "System Events" to key code 19 using control down'`. All I have to do is to get the active space from this command `defaults read com.apple.spaces` and find out the previous or next space ID and press CTRL+1,2,3 based on that. Uh. A lot of work 😅 

How do you get this information from `yabai -m query --spaces`? This JSON is easier to parse. I can also use the `yabai -m query --spaces --space` command to get the current space. Ofc, I can't do this without the yabai service but maybe I can parse the same files as yabai does in my own bash script

**Reply by asmvik** [reply]
yabai is actually building that json using the results of several function calls to request system info.

The active space is determined as follows: 
https://github.com/koekeishiya/yabai/blob/master/src/space_manager.c#L585

Mapping from the internal id to the mission-control index: https://github.com/koekeishiya/yabai/blob/master/src/space_manager.c#L430

**Reply by IonelLupu** [reply]
oh. So I assume I won't be able to get the space from the terminal 😞

**Reply by asmvik** [reply]
I think you would need a tool that does something similar to what I am doing above, unfortunately. I can maybe put together something quick sometime during the weekend if you are unable to find some tool that already does this.

**Reply by asmvik** [reply]
Actually after doing a quick search it looks like https://github.com/davidpurnell/SpaceInfo should do what you need. Not sure if the method it uses to detect the active space works with/without "Displays have separate Spaces" setting, but you can give it a try.

**Reply by IonelLupu** [reply]
Thank you @koekeishiya. I made it work using the SpaceInfo repo you just mentioned and this shell script:
```
declare spaceIdToNumberCode=( ["1"]="18" ["2"]="19" ["3"]="20")

currentSpace=$(~/SpaceInfo --active-space)

if [ "$1" = "next" ]
then
    lastSpace=$(~/SpaceInfo --last-space)
    nextSpace=$(($currentSpace+1))

    if [ $nextSpace -le $lastSpace ]
    then
        osascript -e "tell application \"System Events\" to key code ${spaceIdToNumberCode[$nextSpace]} using control down"
    fi
else
    if [ "$1" = "previous" ]
    then
        firstSpace=$(~/SpaceInfo --first-space)
        previousSpace=$(($currentSpace-1))
        if [ "$1" = "previous" -a $previousSpace -ge $firstSpace ]
        then
            osascript -e "tell application \"System Events\" to key code ${spaceIdToNumberCode[$previousSpace]} using control down"
        fi
    fi
fi
```

It works as expected. It has a slight delay of ~150ms or so until it runs the command but it better than nothing. The osascript takes this much time. This almost makes me go back to Windows 😂 

Thank you one more time for your time! 🚀 ❤️
